### PR TITLE
Fix: `lines-between-class-members` (fixes #9665)

### LINKED ISSUE #9665: False positive on rule `lines-between-class-memebers` if a comment occurs between class members
**Tell us about your environment**

* **ESLint Version:** 4.12
* **Node Version:** 8.9.1
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
n/a

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "lines-between-class-members": 2
    },
    "env": {}
}
```
(generated by [ESLint Demo](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgLyogZXNsaW50LWRpc2FibGUgY2xhc3MtbWV0aG9kcy11c2UtdGhpcyAqL1xuICAvKipcbiAgICogQHJldHVybnMge0FycmF5fVxuICAgKi9cbiAgZ2V0Rm9vKCkge1xuICAgIHRocm93IG5ldyBFcnJvcignQ2hpbGQgY2xhc3Mgc2hvdWxkIHJlaW1wbGVtZW50IHRoaXMgbWV0aG9kJyk7XG4gIH1cbiAgLyogZXNsaW50LWVuYWJsZSBjbGFzcy1tZXRob2RzLXVzZS10aGlzICovXG5cbiAgLyoqXG4gICAqIEByZXR1cm5zIHt1bmRlZmluZWR9XG4gICAqL1xuICBkb1NvbWV0aGluZygpIHt9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImxpbmVzLWJldHdlZW4tY2xhc3MtbWVtYmVycyI6Mn0sImVudiI6e319fQ==))

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I have a class which purpose is to be extended by other classes in order to provide actual implementation of some methods.
The class has placeholder methods that only throw, but don't use context (the subclass implementation will)

Since this is an exception i wrapped the method with `/* eslint-disable class-methods-use-this */` just before the doc block and `/* eslint-enable class-methods-use-this */` just after the method body like that:

```js
class Foo {
  /* eslint-disable class-methods-use-this */
  /**
   * @returns {Array}
   */
  getFoo() {
    throw new Error('Child class should reimplement this method');
  }
  /* eslint-enable class-methods-use-this */

  /**
   * @returns {undefined}
   */
  doSomething() {}
}
```

[Demo showing false positive (demo1)](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgLyogZXNsaW50LWRpc2FibGUgY2xhc3MtbWV0aG9kcy11c2UtdGhpcyAqL1xuICAvKipcbiAgICogQHJldHVybnMge0FycmF5fVxuICAgKi9cbiAgZ2V0Rm9vKCkge1xuICAgIHRocm93IG5ldyBFcnJvcignQ2hpbGQgY2xhc3Mgc2hvdWxkIHJlaW1wbGVtZW50IHRoaXMgbWV0aG9kJyk7XG4gIH1cbiAgLyogZXNsaW50LWVuYWJsZSBjbGFzcy1tZXRob2RzLXVzZS10aGlzICovXG5cbiAgLyoqXG4gICAqIEByZXR1cm5zIHt1bmRlZmluZWR9XG4gICAqL1xuICBkb1NvbWV0aGluZygpIHt9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImxpbmVzLWJldHdlZW4tY2xhc3MtbWVtYmVycyI6Mn0sImVudiI6e319fQ==)

[Demo reduced to bare minimum showing false positive (demo2)](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgZm9vKCkge31cbiAgLyogZm9vICovXG5cbiAgYmFyKCkge31cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibGluZXMtYmV0d2Vlbi1jbGFzcy1tZW1iZXJzIjoyfSwiZW52Ijp7fX19). This one proves that any block comment between class members break the rule

What's interesting: Moving the `/* eslint-enable class-methods-use-this */` from line 9 to line 10 (so that the padding line is above the `/* eslint-enable */` directive) solves the problem (another demo with no error: [demo1-no-error](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgLyogZXNsaW50LWRpc2FibGUgY2xhc3MtbWV0aG9kcy11c2UtdGhpcyAqL1xuICAvKipcbiAgICogQHJldHVybnMge0FycmF5fVxuICAgKi9cbiAgZ2V0Rm9vKCkge1xuICAgIHRocm93IG5ldyBFcnJvcignQ2hpbGQgY2xhc3Mgc2hvdWxkIHJlaW1wbGVtZW50IHRoaXMgbWV0aG9kJyk7XG4gIH1cblxuICAvKiBlc2xpbnQtZW5hYmxlIGNsYXNzLW1ldGhvZHMtdXNlLXRoaXMgKi9cbiAgLyoqXG4gICAqIEByZXR1cm5zIHt1bmRlZmluZWR9XG4gICAqL1xuICBkb1NvbWV0aGluZygpIHt9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImxpbmVzLWJldHdlZW4tY2xhc3MtbWVtYmVycyI6Mn0sImVudiI6e319fQ==), and [demo2-no-error](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgZm9vKCkge31cblxuICAvKiBmb28gKi9cbiAgYmFyKCkge31cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibGluZXMtYmV0d2Vlbi1jbGFzcy1tZW1iZXJzIjoyfSwiZW52Ijp7fX19))

**What did you expect to happen?**
I think the rule should not report an error on line 14, especially since there's a blank line 10 which logically meets rule requirements.
Placement of `/* eslint-* */` directives should not affect whether the rule reports error or not if there's  one empty line between members.

**What actually happened? Please include the actual, raw output from ESLint.**
Rule reports error in line 14, column 3 (see [demo1](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgLyogZXNsaW50LWRpc2FibGUgY2xhc3MtbWV0aG9kcy11c2UtdGhpcyAqL1xuICAvKipcbiAgICogQHJldHVybnMge0FycmF5fVxuICAgKi9cbiAgZ2V0Rm9vKCkge1xuICAgIHRocm93IG5ldyBFcnJvcignQ2hpbGQgY2xhc3Mgc2hvdWxkIHJlaW1wbGVtZW50IHRoaXMgbWV0aG9kJyk7XG4gIH1cbiAgLyogZXNsaW50LWVuYWJsZSBjbGFzcy1tZXRob2RzLXVzZS10aGlzICovXG5cbiAgLyoqXG4gICAqIEByZXR1cm5zIHt1bmRlZmluZWR9XG4gICAqL1xuICBkb1NvbWV0aGluZygpIHt9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImxpbmVzLWJldHdlZW4tY2xhc3MtbWVtYmVycyI6Mn0sImVudiI6e319fQ==) and [demo2](https://eslint.org/demo/#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgZm9vKCkge31cbiAgLyogZm9vICovXG5cbiAgYmFyKCkge31cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibGluZXMtYmV0d2Vlbi1jbGFzcy1tZW1iZXJzIjoyfSwiZW52Ijp7fX19)).

### PR DESCRIPTION
**Pull request by sakabar**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
I fixed the bug of issue #9665 

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

**Comment by sakabar** [comment]
@mysticatea Thank you for your review.
I signed [CLA](https://cla.js.foundation/eslint/eslint?pullRequest=9680). Is this OK?

I added those test cases, which failed. I'm debugging my code now.

**Comment by Aladdin-ADD** [comment]
I see. a blank line was wrongly expected before `/*foo*/` since the rule assumes the comment is for the following member `bar` -- this is not always true.  An easy fix is to check lines after the comment `/*foo*/`.
```js
class Foo{
foo(){} 
/* foo*/

bar(){}
}
```

**Comment by mysticatea** [comment]
Hm, https://github.com/eslint/eslint/pull/9680#issuecomment-348747116 shows "not signed yet", but `licence/cla` check is passed.

@ilyavolodin do you have idea?

**Comment by sakabar** [comment]
@mysticatea Sorry, I did strange things.

1. I committed my code with an old e-mail address. 
2. jsf-clabot worked : not signed yet
3. I noticed that I used the old e-mail address.
4. I did `git reset`, `git commit`, and `git push -f`
5. I signed the CLA with a new e-mail address
6. licence/cla passed. However, the message of jsf-clabot didn't be edited.

**Comment by mysticatea** [comment]
@sakabar Thank you for the information!

Looks like a bug of @jsf-clabot.

**Comment by sakabar** [comment]
@Aladdin-ADD 
> An easy fix is to also check lines after the comment `/*foo*/`.
Sorry, I can't imagine code when there are several comments. I implemented this code in my way.

File: lib/rules/lines-between-class-members.js [hunk 1]
```diff
@@ -55,7 +55,30 @@ module.exports = {
          * @returns {boolean} True if there is at least a line between the tokens
          */
         function isPaddingBetweenTokens(first, second) {
```
Comment (Aladdin-ADD) [inline lib/rules/lines-between-class-members.js]: the function seems a little complicated for me, could you add a comment to say what was happening, thanks!

Comment (sakabar) [reply lib/rules/lines-between-class-members.js]: @Aladdin-ADD Sure, I'll add comments.

**Comment by sakabar** [comment]
@mysticatea 
I added the first and second test case in your review. The third has already been added in [lines-between-class-members.js#L37](https://github.com/eslint/eslint/blob/v4.12.1/tests/lib/rules/lines-between-class-members.js#L37).

All checks have passed.  I'll add more comments because my code is complicated.

**Comment by sakabar** [comment]
@Aladdin-ADD @mysticatea 
Thank you for reviewing. Should I rebase these commits?

**Comment by Aladdin-ADD** [comment]
@sakabar not a necessary, we will squash the commits when merging. 😄

**Comment by sakabar** [comment]
@Aladdin-ADD Sure!

**Comment by mysticatea** [comment]
@platinumazure I think additional invalid tests are unnecessary because this PR is a bug fix of false positive.

**Comment by Aladdin-ADD** [comment]
thanks for contributing! @sakabar

**Comment by sakabar** [comment]
@mysticatea @Aladdin-ADD Thank you for your reviewing!

**Pull request merged**